# Linear Regression Model

In [28]:
import sys, os
src_abspath = os.path.abspath(os.path.join(os.path.split(os.getcwd())[0], 'src'))
sys.path.append(src_abspath)
from db import *
from clf import *

In [40]:
am_debugging = True
if am_debugging:
    np.random.seed(2)
else:
    np.random.seed()

## Data/Features
#### Composite Dataset

In [32]:
data = feat_iculos().iloc[:500]

#### Partition: Training & Testing Sets

In [36]:
X_train, X_test, y_train, y_test = data_partition(data)

X_train.shape, X_test.shape, y_train.shape, y_test.shape

((350, 16), (150, 16), (350,), (150,))

#### Normalization

In [37]:
scaled_X_train, scaled_X_test = scale(X_train, X_test)

#### Feature Selection

In [60]:
pass

## Model: Supervised Learning

In [38]:
clf_model, y_pred, best_p, best_score = gridsearch(scaled_X_train, scaled_X_test, y_train)
print "Best Parameters: ", best_p
print "Best Grid Search Score: ", best_score
print "Best Estimator: ", clf_model, "\n"

Best Parameters:  {'normalize': True, 'fit_intercept': True}
Best Grid Search Score:  -0.0409464857579
Best Estimator:  LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=True) 



#### Compare Predictions
Compare predictions of last 10 data points against the true values (`y_test`).

In [41]:
y_pred[-10:]

array([  94.67467893,   41.15713512,  106.17360746,   56.33360743,
         48.16493186,  132.57932916,  158.06642203,   79.17522761,
        125.12089252,   94.19920334])

In [42]:
y_test[-10:]

array([ 182.83,   48.12,   46.09,   26.63,   54.03,   52.28,   70.41,
         95.  ,  358.92,   12.44])

### Metrics

#### Root Mean Squared Error (RMSE)
Indicates the *absolute fit* of the model to the data. In other words, how close the observed data points are to the model's predicted values.

Also, it is the standard deviation of unexplained variance (residuals).

In [57]:
rmse = RMSE(y_pred, y_test)
print "RMSE = %.3f days" % rmse

RMSE = 99.508 days


In [56]:
print "[%.2f,%.2f]" % (y_test.min(), y_test.max())

[5.16,588.23]


The linear regression model has an RMSE value of 99.508 days. The value is somewhat small relative to the range of the `dependent variable` (average ICU LOS). Thus, indicating that model has decent accuracy in predicting the response.

#### Mean Absolute Error
The Mean Absolute Error measures how close the model's predictions are to the observed values.

In [59]:
mae = MAE(y_pred, y_test)
print "MAE = %.3f days" % mae

MAE = 66.779 days


The average difference between prediction and observation is 66.8 days, which is high.

#### R<sup>2</sup>: Coefficient of Determination
R<sup>2</sup> quantifies the goodness of fit of the linear model. More specifically, it depicts the predictive power of the model.

Range: [0, 1]

In [54]:
r2(y_pred, y_test)

0.017654572548759906

The R<sup>2</sup> value is approximately 0.02, which is low and indicates that the model does not have strong predictive power.